0065-drop-dunder-utils.md

```diff
@@ -85,20 +101,48 @@ There are some things which we just can't control, see:
 ## Unresolved questions
 [unresolved]: #unresolved-questions
 
-None so far.
+### Many others may also be using ``salt://_utils/``. That can't be replaced with a simple import. That code would need to be packaged and installed separately into the minion's environment. — By [James Howe](https://github.com/OrangeDog)
+
+What exactly would be a use case for the ``salt://_utils/`` usage?
+
 
 ## Resolved questions
 [resolved]: #resolved-questions
 
-### Is this loader accessible anywhere else? e.g. via utils in templates? That significantly increases the surface of the breaking change. But that's not necessarily a bad thing.
+### Is this loader accessible anywhere else? e.g. via utils in templates? That significantly increases the surface of the breaking change. But that's not necessarily a bad thing. — By [James Howe](https://github.com/OrangeDog)
 
 Thank god no.
 
-### Breaking change on public API for custom modules (and extension modules?)
+### Breaking change on public API for custom modules (and extension modules?) — By [James Howe](https://github.com/OrangeDog)
 
 During the deprecation period, ``__utils__`` shall be available, although a warning about it's
 deprecation will be emmitted.
 
+### But still, is there a cleaner way to solve the original use-case (distribute common code, or override built-in utils)? — By [Max Arnold](https://github.com/max-arnold)
+
+Distributing common code has been solved in the python ecosystem, ie, packaged libraries.
+As for overriding built-in utils, this is one of the major reasons why the utils loader should have never been implemented.
+One thing is being able to override an execution module, or state module, the exposure surface is small, as for
+utility code, the surface is unmeasureable because that code is meant to be reused accross the whole code base. One small change
+in the overridden code has the potential of bringing a Salt installation to the ground.
+
+### I guess shipping custom utils through salt extensions will be also broken? — By [Max Arnold](https://github.com/max-arnold)
+
+A salt extension is a packaged python library that Salt can use, so, why not just import the utilitary code and just use it.
+
+### I do not mind to remove ``__utils__``, but what mechanism should I use now? — By [Alberto Planas](https://github.com/aplanas)
+
+Packaged libraries distributed to the minions. Salt can even distribute and install these on the minions.
+
+### For example, there are shared code that I do not want to be represented as a module. — By [Alberto Planas](https://github.com/aplanas)
+
+I don't know enough about this particular usage or it's reasonings, so, why not?
+
+### Will any Python submodule inside a salt module be recollected and send to the minion? — By [Alberto Planas](https://github.com/aplanas)
+
+I consider this a feature request.
+Although it could relate to this SEP, its work/implementation should be separate.
```

Comment:
> I agree. But at least this RFC should order both to guarantee that there is an alternative before the removal.

To me, there is, package your utils code.


I keep mentioning that this is **my** personal view on the issue, a harsh one because I think ``__utils__`` should have never been implemented.

But do note that, I'm not enough to approve this SEP, and a few of my teammates might not even share this harsh perspective on this subject.

I **do** value your feedback, and even though it hasn't moved my position, it could move someone else's position on the core team.
This SEP can still be rejected, or forced to change to avoid being rejected.

Reply:
Uhmm so many **bold** text makes me wonder if there is a miscommunication here.

I am not opposed to the idea. IMHO all `__XXXX__` is an undesirable hack, I do not see `__utils__` special here.

> I think __utils__ should have never been implemented.

Even tho I do not understand the specific issue about `__utils__`, I trust your opinion here.

I make an effort to go to some of the issues linked in the document, to try to figure out what is so bad specifically for `__utils__`, but I only saw issues in the modules and in how the generic loader works (that is also what I see problematic).  Maybe clarifying the specifics about `__utils__` in the RFC can help to gain fast agreement.

Reply:
> Uhmm so many bold text makes me wonder if there is a miscommunication here.

Sorry for the bold :grinning:
I'm just trying to state that this is not, at least yet, the opinion of the majority of core team, just me.

> I am not opposed to the idea. IMHO all ``__XXXX__`` is an undesirable hack, I do not see ``__utils__`` special here.

Yes, it is. I also have ideas about all of that, but I might not even propose a SEP for that, it touches TOO much, and any deprecation path or compat layer might just be too complicated to implement.

> I make an effort to go to some of the issues linked in the document, to try to figure out what is so bad specifically for __utils__, but I only saw issues in the modules and in how the generic loader works (that is also what I see problematic). Maybe clarifying the specifics about __utils__ in the RFC can help to gain fast agreement.

* https://github.com/saltstack/salt/pull/62006
* https://github.com/saltstack/salt/pull/62007
* https://github.com/saltstack/salt/pull/62021

The loader is mean to "store"/"hold" functions(as simple as they can be), which provide user facing functionality.
Utilities are not user facing functionality, that why I say ``__utils__`` should have never been implemented. But it did.
On and off we issues related to ``__utils__``, the most recent ones are related by memory consumption. In sum, because the salt loaders load the same module over and over again, some utility modules relying on the `warnings` or `atexit` modules, keep adding to some internal lists of those stdlib modules.

This should never happen, and it doesn't in user facing loader modules, which, ``__utils__`` is not.

Reply:
When comparing alternative ways to achieve the same need, we need to consider the complexity/usability as well.

Use-case 1: Ship reusable code that can be called from execution modules to minions
Constraint 1: The code should not be callable directly from Salt CLI
Constraint 2: Private functions are not exposed by the loader, so it is not possible to place these custom functions into an execution module itself.

Solution 1: `_utils` folder and sync_all
Solution 2: Package as a Python module. Downsides: lower development speed, need to mess with Python packaging, pollute PyPI with internal packages or maintain your own PyPI/Devpi server, additional states to install it (as opposed to a sync_all)

Solution 3 (idea): untangle the `utils` module namespace and `__utils__`.  Allow the latter to invoke only user supplied modules shipped in `_utils` and not Salt core utils.
Solution 4 (idea): implement some convention to cross-call private execution functions without exposing them to users via CLI and docs

Use-case 2: Override built-in utils.
Why 1: patch an issue that is not yet fixed and released
Why 2: can't upgrade Salt yet, but need a feature/fix ASAP

Solution 1: Override an utility module through `_utils`. Downsides: sometimes doesn't work, because Salt often imports utils instead of using `__utils__` everywhere. Solution: fix Salt to always use `__utils__`
Solution 2: Dramatically speed up bug fixes and release process. This was the original goal of the development process overhaul several years ago (I think a Salt core dev even hinted that it would be possible to achieve monthly/weekly releases with some automation). Apparently, it is no longer a goal (see the LTS SEP)
Solution 3: Significantly simplify the Salt upgrade process. Maybe via Tiamat packages? Ship some built-in modules that can reliably automate that. Significantly decrease number of regressions in Salt releases and improve compatibility, so upgrading is safe and painless

Reply:
> When comparing alternative ways to achieve the same need, we need to consider the complexity/usability as well.

> **Use-case 1**: Ship reusable code that can be called from execution modules to minions
> **Constraint 1**: The code should not be callable directly from Salt CLI
> **Constraint 2**: Private functions are not exposed by the loader, so it is not possible to place these custom functions into an execution module itself.

It is, but you'd need to import them. To me this would be ok since it's not on the Salt side, although another hack/work-around, so, not the best solution.


> **Solution 1**: `_utils` folder and sync_all

This is what we have, and reasons were already shown why ``__utils__`` was a bad idea. No one likes to see their server's memory go constantly up.

> **Solution 2**: Package as a Python module. Downsides: lower development speed, need to mess with Python packaging, pollute PyPI with internal packages or maintain your own PyPI/Devpi server, additional states to install it (as opposed to a sync_all)

True, but that's how it should have been from the get go. And, not doing this means the burden is on the Core team, which has hundreds of modules to maintain.
Packaging on the user end, although it adds complexity on the user side, also means the user can release way faster than Salt.

> **Solution 3 (idea)**: untangle the `utils` module namespace and `__utils__`.

Not really a solution, this is where we currently are.

> Allow the latter to invoke only user supplied modules shipped in `_utils` and not Salt core utils.

Not a bad idea. At least no one would be able to override Salt's internal utils and it's usage. Any bugs or memory consumption issue would be on the user's end.
Although, we all know that the Salt Project would get the bug reports, so, more triaging to do...

> **Solution 4 (idea)**: implement some convention to cross-call private execution functions without exposing them to users via CLI and docs

Hmm, might be a tad complicated logic to implement and we don't get rid of ``__utils__``.


> **Use-case 2**: Override built-in utils.
> **Why 1**: patch an issue that is not yet fixed and released
> **Why 2**: can't upgrade Salt yet, but need a feature/fix ASAP
> 
> **Solution 1**: Override an utility module through `_utils`.
> Downsides: sometimes doesn't work, because Salt often imports utils instead of using `__utils__` everywhere.
> Solution: fix Salt to always use `__utils__`

I'm strongly against this, as we can see from the proposed SEP.

> **Solution 2**: Make a custom Salt package. Involves a lot of work to maintain and some infrastructure to host

True, but if not like this, then the Core team has the burden, and it's not just one or two modules to maintain, it's ALL of them.

> **Solution 3**: Dramatically speed up bug fixes and release process. This was the original goal of the development process overhaul several years ago (I think a Salt core dev even hinted that it would be possible to achieve monthly/weekly releases with some automation). Apparently, it is no longer a goal (see the LTS SEP) 

This is not true. The goal remains, I can even say that our optimal solution would be nightly builds for those who want to use them.
The LTS is for those that don't want to live on the edge or have higher constraints to upgrading Salt.
ALL bug and security fixes go into the LTS and would be released as a minor version release.

> **Solution 4**: Significantly simplify the Salt upgrade process. Maybe via Tiamat packages? Ship some built-in modules that can reliably automate that. Significantly decrease number of regressions in Salt releases and improve compatibility, so upgrading is safe and painless

This is also our goal, Tiamat packages and Significantly decrease the number of regressions.
As you can imagine though, right now, it's a lot of code to maintain, so, even though the Core team is making progress, we're not there yet.

Reply:
> > **Solution 1**: `_utils` folder and sync_all
> 
> This is what we have, and reasons were already shown why `__utils__` was a bad idea. No one likes to see their server's memory go constantly up.

I can be wrong, but I interpreted this proposal as dropping `__utils__` as a load mechanism, but maintain `_utils` sync_all synchronization as a deliver mechanism for plain Python code.  This code is later accessible via normal "import" (incidentally this is how I am using it).

Reply:
> > > **Solution 1**: `_utils` folder and sync_all
> > 
> > 
> > This is what we have, and reasons were already shown why `__utils__` was a bad idea. No one likes to see their server's memory go constantly up.
> 
> I can be wrong, but I interpreted this proposal as dropping `__utils__` as a load mechanism, but maintain `_utils` sync_all synchronization as a deliver mechanism for plain Python code. This code is later accessible via normal "import" (incidentally this is how I am using it).

I'm not against the delivery of ``_utils`` if we still drop ``__utils__``

Reply:
> I'm not against the delivery of `_utils` if we still drop `__utils__`

If this is technically OK, I think that this can be a nice compromise. We drop some use cases for `__utils__`, but there is still a clear way to deliver user Python code that is required for modules and states.

Reply:
> > I'm not against the delivery of `_utils` if we still drop `__utils__`
> 
> If this is technically OK, I think that this can be a nice compromise. We drop some use cases for `__utils__`, but there is still a clear way to deliver user Python code that is required for modules and states.

I'm not against `_utils` because it's not adding burden to Salt and it's maintenance, ``__utils__`` is.
We'd need to clearly document how to use modules delivered from `_utils` though and perhaps provide use cases.

I'd need to check how this would play out, but, I could even consider leaving ``__utils__`` for modules shipped through ``_utils``, but preferably, we would just get rid of all `__utils__` usage to avoid confusion.

Reply:
> how to use modules delivered from `_utils`

I'm pretty sure the only way it will work is via a loader, which for consistency should be named `__utils___`. It shouldn't have behaviour that's different from the other synced loaders.

Reply:
> > pollute PyPI with internal packages or maintain your own PyPI/Devpi server
> 
> Or just use a different installation source. `pip` can install from git, an archive file (including `.whl`), a source folder, etc,.

Right.


> > **Solution 3** (idea): untangle the `utils` module namespace and `__utils__`. Allow the latter to invoke only user supplied modules shipped in `_utils` and not Salt core utils.
> 
> That sounds like a good idea.

No, still does not sound like a good idea, the memory issues from the stdlib won't just go away and we shouldn't constraint how we write utility functions, we already do that for all of the other loaders...
What would be the alternative, utility functions for the utils loader?!

> 
> > how to use modules delivered from `_utils`
> 
> I'm pretty sure the only way it will work is via a loader, which for consistency should be named `__utils___`. It shouldn't have behaviour that's different from the other synced loaders. Otherwise the reload wouldn't work, right?

Hmm, true. I could probably compromise on keeping ``__utils__`` which only has access to whatever is provided from ``_utils``, though, again, the Salt project will likely still have issues filed because of memory consumption, although, no longer directly related to Salt itself.

Reply:
> Keeping a *.whl binary in a git repo is not a good practice

Which is why I didn't suggest it. It can be anywhere you like - local file, salt fileserver, a web server.